File: src/com/github/gtache/client/ProcessOverSocketStreamConnectionProvider.scala
/** *****************************************************************************
  * Copyright (c) 2017 Rogue Wave Software Inc. and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-v10.html
  *
  * Contributors:
  * Michał Niewrzał (Rogue Wave Software Inc.) - initial implementation
  * ******************************************************************************/
package com.github.gtache.client

import java.io.{IOException, InputStream, OutputStream}
import java.net.{ServerSocket, Socket}
import java.util.Objects

import com.intellij.openapi.diagnostic.Logger

/**
  * Unused at the moment
  *
  * @param commands
  * @param workingDir
  * @param port
  */
class ProcessOverSocketStreamConnectionProvider(commands: Seq[String], workingDir: String, port: Int = 0) extends ProcessStreamConnectionProvider(commands, workingDir) {
  private val LOG = Logger.getInstance(classOf[ProcessStreamConnectionProvider])
  private var socket: Socket = _
  private var inputStream: InputStream = _
  private var outputStream: OutputStream = _

  @throws[IOException]
  override def start(): Unit = {
    val serverSocket = new ServerSocket(port)
    val socketThread = new Thread(() => {
      try
        socket = serverSocket.accept
      catch {
        case e: IOException =>
          LOG.error(e)
      } finally try
        serverSocket.close()
      catch {
        case e: IOException =>
          LOG.error(e)
      }
    })
    socketThread.start()
    super.start()
    try {
      socketThread.join(5000)
    }
    catch {
      case e: InterruptedException =>
        LOG.error(e)
    }
    if (socket == null) throw new IOException("Unable to make socket connection: " + toString) //$NON-NLS-1$
    inputStream = socket.getInputStream
    outputStream = socket.getOutputStream
  }

  override def getInputStream: InputStream = inputStream

  override def getOutputStream: OutputStream = outputStream

  override def stop(): Unit = {
    super.stop()
    if (socket != null) try
      socket.close()
    catch {
      case e: IOException =>
        LOG.error(e)
    }
  }

  override def hashCode: Int = {
    val result = super.hashCode
    result ^ Objects.hashCode(this.port)
  }
}

File: intellij-lsp/src/main/kotlin/com/github/gtache/lsp/client/connection/ProcessOverSocketStreamConnectionProvider.kt
/* Adapted from lsp4e */
package com.github.gtache.lsp.client.connection

import com.intellij.openapi.diagnostic.Logger
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.net.ServerSocket
import java.net.Socket
import java.util.*
import kotlin.concurrent.thread

/**
 * A connection to a process using sockets
 *
 * @param commands
 * @param workingDir
 * @param port
 */
class ProcessOverSocketStreamConnectionProvider(commands: Array<String>, workingDir: String, private val port: Int = 0) :
    ProcessStreamConnectionProvider(commands, workingDir) {

    companion object {
        private val logger = Logger.getInstance(ProcessOverSocketStreamConnectionProvider::class.java)
    }

    override var inputStream: InputStream? = null
    override var outputStream: OutputStream? = null
    override val errorStream: InputStream?
        get() = inputStream

    private var socket: Socket? = null

    override fun start(): Unit {
        val serverSocket = ServerSocket(port)
        val socketThread = thread {
            try {
                socket = serverSocket.accept()
            } catch (e: IOException) {
                logger.error(e)
            } finally {
                try {
                    serverSocket.close()
                } catch (e: IOException) {
                    logger.error(e)
                }
            }
        }
        socketThread.start()
        super.start()
        try {
            socketThread.join(5000)
        } catch (e: InterruptedException) {
            logger.error(e)
        }
        if (socket == null) throw IOException("Unable to make socket connection: " + toString()) //$NON-NLS-1$
        inputStream = socket?.getInputStream()
        outputStream = socket?.getOutputStream()
    }

    override fun stop(): Unit {
        super.stop()
        try {
            socket?.close()
        } catch (e: IOException) {
            logger.error(e)
        }
    }

    override fun equals(other: Any?): Boolean {
        return super.equals(other) && other is ProcessOverSocketStreamConnectionProvider && port == other.port
    }

    override fun hashCode(): Int {
        val result = super.hashCode()
        return Objects.hash(result, this.port)
    }
}